src/Tasks/Microsoft.NET.Build.Tasks/WriteAppConfig.cs

```diff
@@ -0,0 +1,171 @@
+                    targetFrameworkIdentifier,
+                    targetFrameworkVersion,
+                    runtimeVersion,
+                    out XElement supportedRuntime))
+                {
+                    if (startupNode == null)
+                    {
+                        startupNode = new XElement("startup");
+                        doc.Root.Add(startupNode);
+                    }
+
+                    startupNode.Add(supportedRuntime);
+                }
+            }
+        }
+
+        private static bool HasExistingSupportedRuntime(XElement startupNode)
+        {
+            return startupNode != null
+                  && startupNode.Nodes().OfType<XElement>().Any(e => e.Name.LocalName == "supportedRuntime");
+        }
+
+        //https://github.com/dotnet/docs/blob/master/docs/framework/configure-apps/file-schema/startup/supportedruntime-element.md
+        private static bool TryGetSupportRuntimeNode(
+            string targetFrameworkIdentifier,
+            string targetFrameworkVersion,
+            string runtimeVersion,
+            out XElement supportedRuntime)
+        {
+            supportedRuntime = null;
+
+            if (targetFrameworkIdentifier == ".NETFramework"
+                && NuGetVersion.TryParse(targetFrameworkVersion.TrimStart('v', 'V'), out NuGetVersion parsedVersion))
+            {
+                if (parsedVersion.Version.Major < 4)
+                {
+                    string supportedRuntimeVersion = null;
+
+                    if (parsedVersion.Version.Major == 1 && parsedVersion.Version.Minor >= 0 && parsedVersion.Version.Minor > 1)
+                    {
+                        supportedRuntimeVersion = "v1.0.3705";
+                    }
+                    else if (parsedVersion.Version.Major == 1 && parsedVersion.Version.Minor >= 1)
+                    {
+                        supportedRuntimeVersion = "v1.1.4322";
+                    }
+                    else if (parsedVersion.Version.Major >= 2 && parsedVersion.Version.Major < 4)
+                    {
+                        supportedRuntimeVersion = "v2.0.50727";
+                    }
+
+                    if (supportedRuntimeVersion == null)
+                    {
+                        return false;
+                    }
+
+                    supportedRuntime =
+                           new XElement(
+                               "supportedRuntime",
+                               new XAttribute("version", supportedRuntimeVersion));
+
+                    return true;
+                }
+                else if (parsedVersion.Version.Major == 4)
+                {
+                    supportedRuntime =
+                        new XElement(
+                            "supportedRuntime",
+                            new XAttribute("version", "v4.0"),
+                                new XAttribute("sku", $"{targetFrameworkIdentifier},Version={targetFrameworkVersion}"));
```

Comment:
For completeness, sku should include Profile (if any) as well in TFM. We don't support net40-client directly with our target framework inference, but I believe some folks do it explicitly or use community "extras" package to do it.

Reply:
an example? And is there a easy way to test it to make sure profile and tfm is in correct "format"? I am afraid of adding more property and it ends up cannot be read by runtime. Today, I avoid this issue by matching exactly to the following doc.

https://github.com/dotnet/docs/blob/master/docs/framework/configure-apps/file-schema/startup/supportedruntime-element.md

Reply:
I find it, it is also a doc

https://docs.microsoft.com/en-us/dotnet/framework/configure-apps/file-schema/startup/supportedruntime-element

So it will just do Profile="${TargetFrameworkProfile}"

Reply:
You will need to make profile optional and only use it if set.

Reply:
Fixed, it is optional. Only add it when it is set